### PR TITLE
Add `ArrayBuffer.isView` + fixes for `Array.isArray`

### PR DESCRIPTION
**Pull request by MaxGraey**
- ArrayBuffer.isView
- plus other minor additions

**EDIT**
About [ArrayBuffer.isView() on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/isView)

### CODE REVIEW

File: std/portable/index.js [hunk 1]
```diff
@@ -197,15 +204,16 @@ globalScope["isFloat"] = function isFloat(arg) {
   return typeof arg === "number";
 };
 
-globalScope["isReference"] = function isClass(arg) {
-  return typeof arg === "object" || typeof arg === "string";
+globalScope["isReference"] = function isReference(arg) {
+  return typeof arg === "object" || typeof arg === "string" || typeof arg === "symbol";
```
Comment (dcodeIO) [inline std/portable/index.js]: Not sure about symbols here. Not really a pointer, and can't really `load` from it.

Comment (MaxGraey) [reply std/portable/index.js]: got it. I'll remove it

**Comment by MaxGraey** [comment]
`ArrayBuffer.isView` has more general behaviour. It check any typed array and DataView in one call. So if you need verify some arg is any typed array via `instanceof` it looks like:
```ts
if (
  arg instanceof UInt8Array || 
  arg instanceof Int8Array   || 
  arg instanceof Int32Array || 
  ...
  arg instanceof DataView
) {
   ...
}
```
instead:
```ts
if (ArrayBuffer.isView(arg)) {
  ...
}
```

**Comment by MaxGraey** [comment]
The main reason for having builtin `isArrayBufferView` is faster compiling I guess. WDYT?

**Comment by dcodeIO** [comment]
One issue I see with the proposed solution is that it checks a class structurally which can, technically, lead to false positives. If there is a need for this now, I'd suggest checking whether the type extends `TypedArray<T>` or `DataView` in that builtin to avoid any issues. Essentially, get a reference to those classes on initialization of the program, and check them in the builtin.

**Comment by MaxGraey** [comment]
Yeah, but I considering `isArray` and `isArrayBufferView` as structural checking so it probably better call `isArrayLike` and `isArrayBufferViewLike`. Also will be great verify member's type

**Comment by MaxGraey** [comment]
Also this fix false positive issue when `isArray` accept strings as arrays:
```ts
const str = "hello";
assert(isArray(str) == true); // valid
```

**Comment by dcodeIO** [comment]
Note that structural comparison as proposed can still yield false positives. Also note that `isArray` is not `Array.isArray` and is meant to [check whether a value can be used like an array](https://github.com/AssemblyScript/assemblyscript/blob/master/std/assembly/index.d.ts#L131).

I agree though that `isArray` alone is probably not enough to implement `Array.isArray` properly. What we'd need is a way to test whether a type **actually** implements any `Array<T>`, but that's not `isArray` as it stands.

**Comment by MaxGraey** [comment]
I suggest add `isArrayLike<T>(arg: T)` which actually will be less strict

**Comment by dcodeIO** [comment]
So, to recap, `isArray` would have to check, as a built-in, whether a type actually extends `Array<T>` (but not `TypedArray<T>`, and not structurally), while `isArrayLike` would be what `isArray` is now. That's about your thinking?

**Comment by MaxGraey** [comment]
yes!

**Comment by dcodeIO** [comment]
Ok, agreeing so far, might also want a `isTypedArray<T>` builtin then for completeness I think.

**Comment by MaxGraey** [comment]
Also this will be pretty useful for `Array.from` which should allow in future this
```ts
Array.from({ length: 100 }, val => val)
```

**Comment by MaxGraey** [comment]
> Ok, agreeing so far, might also want a isTypedArray<T> builtin then for completeness I think.

You mean `isTypedArrayLike<T>`? 

What is preferable way? Add all this in current PR or do it in separate one?

**Comment by dcodeIO** [comment]
> You mean isTypedArrayLike<T>?

No, I meant:

* isArray<T> -> checks whether `T` **actually** is or extends any `Array<*>`
* isTypedArray<T> -> checks whether `T` **actually** is or extends any `TypedArray<*>`
* isArrayLike<T> -> checks whether `T` allows indexed access and has a `.length`

With that we should be able to distinguish all the cases.

> What is preferable way? Add all this in current PR or do it in separate one?

Making it a separate PR makes sense to me, yeah.

**Comment by MaxGraey** [comment]
Got it. So I create separate PR for this after merging this if you don't mind

**Comment by MaxGraey** [comment]
Synced with master

File: tests/compiler/std/array.ts [hunk 2]
```diff
@@ -33,11 +33,13 @@ var arr = new Array<i32>();
 class P {}
 var num = 1;
 var Null: i32[] | null = null;
+var str = "abc";
 assert(Array.isArray(Null) == false);
 assert(Array.isArray(arr) == true);
 assert(Array.isArray(new P()) == false);
-// assert(Array.isArray(new Uint8Array(1)) == false); fail
```
Comment (dcodeIO) [inline tests/compiler/std/array.ts]: seems to be left over

File: src/builtins.ts [hunk 3]
```diff
@@ -134,9 +133,13 @@ export function compileCall(
...
-      return classType !== null && classType.lookupOverload(OperatorKind.INDEXED_GET) !== null
-        ? module.createI32(1)
-        : module.createI32(0);
+      return (
+        classType !== null &&
+        classType.lookupField("buffer_") !== null &&
+        classType.lookupField("length_") !== null &&
+        classType.lookupOverload(OperatorKind.INDEXED_GET) !== null &&
+        classType.lookupOverload(OperatorKind.INDEXED_SET) !== null
+      ) ? module.createI32(1) : module.createI32(0);
```
Comment (dcodeIO) [inline src/builtins.ts]: Can we leave this for a later PR? Otherwise, it should check whether `Array<*>` is in the prototype chain.

Comment (MaxGraey) [reply src/builtins.ts]: Sure

Comment (MaxGraey) [reply src/builtins.ts]: Done!

Comment (dcodeIO) [reply src/builtins.ts]: To bubble up all levels, for example in an `A extends B extends C extends Array<T>` case:

```ts
var extendsArray = false;
var current = classType;
do {
  if (current.prototype == program.arrayPrototype) {
    extendsArray = true;
    break;
  }
} while (current = current.base);
```

Comment (MaxGraey) [reply src/builtins.ts]: I think make sense create function for that and called it like `hasPrototypeOf(proto1, proto2)`. WDYT?

Comment (MaxGraey) [reply src/builtins.ts]: I found this: https://github.com/AssemblyScript/assemblyscript/blob/2131c519321ecfb0b1c1a642ba9d36f3c18a81cc/src/program.ts#L2937

Comment (dcodeIO) [reply src/builtins.ts]: Sure, makes sense. Maybe `ClassPrototype#extends`, so one can do `if (classType.prototype.extends(program.arrayPrototype)) { ... }`?

Comment (dcodeIO) [reply src/builtins.ts]: Yeah, though `isAssignableTo` does this on concrete types (resolved generic arguments), not just prototypes (unresolved generic arguments). That's the check when assigning a reference to a reference var.

File: src/program.ts [hunk 4]
```diff
@@ -2856,6 +2856,15 @@ export class ClassPrototype extends Element {
     this.decoratorFlags = decoratorFlags;
   }
 
+  extends(basePtototype: ClassPrototype | null): bool {
+    if (!basePtototype) return false;
```
Comment (dcodeIO) [inline src/program.ts]: Should already be covered by the first `current == basePrototype` of the loop

Comment (MaxGraey) [reply src/program.ts]: right!

**Comment by dcodeIO** [comment]
Thanks! :)

**Pull request merged**